Sorting is just Character::pickup() and Character::drop(). How hard could it be?

RenechCDDA opened the pull request:
#### Summary
Features "Actual moving to your sorting destination! With a bag full of items!"

#### Purpose of change
* Alternative to #84297

"Actually" pick stuff up and put it down.

#### Describe the solution
Don't stop picking things up until we're full of things to sort or all the items to be sorted aren't going to the same place

Drop things at the destination until we're out of things to drop

#### Describe alternatives you've considered


#### Testing

https://github.com/user-attachments/assets/4f856c59-92fe-4322-bc98-354881961ab6



#### Additional context

I have severe, incredible criticisms of the code I've written here. It is confusing, loops in counterintuitive ways, and overall is really not great.

Known issues:
Many. This *will* need followup PRs.

Known solved issues:
It moves entire bags worth of stuff at once. Congratulations, me.

RenechCDDA committed:
Rought draft?

RenechCDDA committed:
This doesn't work?

RenechCDDA committed:
Pick stuff up, put stuff down

Reverts test changes from https://github.com/CleverRaven/Cataclysm-DDA/pull/62723

...Those changes appear to test that the *loot sorting* option successfully unloads (?) items. As far as I can ask, that has never been the case.

I do not know how this test was passing before, but it was.

After some consultation with other devs on the devcord, we couldn't find a reason for it existing, so simply removing the erroneous fail was the decided upon course of action.

past-simple commented:
> Known issues:
> Many. This will need followup PRs.

Gently and with sympathy, since everyone working on the game has to share master, wouldn't it make sense to keep this in its own branch until the issues were fixed? This would be very normal in the software development world (and the open source world).

Alternatively, it was sounding like there might be other ideas to fix testing zone walkability that didn't involve full simulation of the transfer and were less likely to introduce huge side effects? Those might be less work overall for the developer and the computer as well.

RenechCDDA commented:
> > Known issues:
> > Many. This will need followup PRs.
> 
> Gently and with sympathy, since everyone working on the game has to share master, wouldn't it make sense to keep this in its own branch until the issues were fixed? This would be very normal in the software development world (and the open source world).
> 
> Alternatively, it was sounding like there might be other ideas to fix testing zone walkability that didn't involve full simulation of the transfer and were less likely to introduce huge side effects? Those might be less work overall for the developer and the computer as well.

And who's going to test on that branch? Just me? Maybe if I plug 20 more hours into it I'll have a revelation? 

Who's going to PR on that branch? Just me, again?

If you don't want experimental, don't play experimental. Sometimes things are going to come in even when we know they're not 100%. You're free to argue for how close to 100% they should be, but asking for *stable* levels of commitment on the experimental branch defeats the entire purpose, and puts an enormous burden on individual contributors.

Your argument is that you don't want to play it. Don't, then. You can't appeal to the idea of "the people working on it"... because I am one of those people, I speak with almost all of our regular contributors everyday. I cannot think of a single one that would agree that PRing feature changes is ever going to make it harder to *work on* the game, because that just doesn't make any sense.

ShnitzelX2 reviewed (changes requested):
The overall plan here is good, please see my review comment and I'll review this again after the necessary changes are made.

ShnitzelX2 commented on src/activity_actor.cpp:
```diff
@@ -10303,37 +10340,90 @@ void zone_sort_activity_actor::stage_do( player_activity &act, Character &you )
-                continue; // Not a valid destination
+        // Picking up
+
+        if( !you.activity.targets.empty() ) {
```
`player_activity` is the class that still does the general activity handling. In this case, `zone_sort_activity_actor` is a pointer held in `player_activity::actor`, for migration purposes.

`player_activity::do_turn()` calls in this case `zone_sort_activity_actor::do_turn()` which eventually runs `zone_sort_activity_actor::stage_do()`.

`player_activity &act` ends up being a reference to the `player_activity` the actor is owned by. So `you.activity` will be the same for a currently running activity as `player_activity &act`. 

_However_, the point is again that e.g. `you.activity.targets` is a legacy member of `player_activity` and should not be used. See line 53 of player_activity.h for all of the legacy members.

RenechCDDA replied:
I believe I've cleared that up. While I was here I also pushed some commits to deal with the worst failure states....

There remains a FIXME notice for if we're standing next to a valid destination but nothing can be dropped. I am unsure if that still needs to be resolved after using put_into_vehicle_or_drop_ret_locs(), it's difficult to test manually.

RenechCDDA replied:
Okay, I think I've got fixes for most of these. And not even all of them were my fault! 🚀 

Here's some saves to verify the behavior.

Shopping cart as loot assistant: 
[SORT_TEST_01_PUT_IN_CART-trimmed.tar.gz](https://github.com/user-attachments/files/24388554/SORT_TEST_01_PUT_IN_CART-trimmed.tar.gz)
This was just a miss on my end.

4x4 truck (behavior appeared to differ from your report, but definitely a bug with the return value!):
[SORT_TEST_02_PUT_IN_VEH-trimmed.tar.gz](https://github.com/user-attachments/files/24388555/SORT_TEST_02_PUT_IN_VEH-trimmed.tar.gz)


Sorting into air (see below as well): 
[SORT_TEST_03_TRY_PUT_IN_AIR-trimmed.tar.gz](https://github.com/user-attachments/files/24388556/SORT_TEST_03_TRY_PUT_IN_AIR-trimmed.tar.gz)


>Sorted some books where the loot tiles were open air. Correctly got the "can't sort onto open air!" message and the player didn't move, but the game froze.

Hmm, I'm not able to replicate on my end, but I put in the missing abort for that scenario, just in case.

>I made a new unsorted zone with five heavy-duty fridges and sorted it into a default zone. This worked, the heavy-duty fridges were moved by wielding one at a time, though one was dropped outside of the zone even though there were available spaces in the zone
> [...]
>I spawned 25 fridges (taking up maximum volume in a 5x5 space) and tried to sort the 1000-book pile onto the center of them. The books disappear when dropped, but this happens with a normal drop as well.

This appears to be due to map::add_item_or_charges()'s overflow behavior... I am not sure of any simple workarounds we can perform here. There's not even a simple way to pre-check if the item will fit at the destination.

>Had both the player and the NPC sort do the same sort. Started okay, but then the player ran into the NPC and cancelled the activity, and when I started sorting again the game froze. Replicating this may be difficult.

mmmmmmm very little clue what this could be. Maybe something to do with stack processing, if one of you takes an item while the other is partway through the stack with a `num_processed` > 0? I would think that would result in "skipped items" at worst though, which would eventually get cleaned up when the stage goes back to THINK.

RenechCDDA committed:
Use local members for targets/coords
Add some more comments while I'm here

RenechCDDA committed:
Add some handling for stuffing items into your inventory

Fix the serialization for the zone_sort_actor members (oops)

RenechCDDA committed:
Fix some of my own FIXMEs

Fix IWYU complaints(???)

RenechCDDA commented:
Sorry for reviewers/future git blamers. These commits are a mess.

RenechCDDA committed:
Fix some sorting bugs

-Actually return an item location in try_to_put_into_vehicle() if we put something into the vehicle!!!

-Properly take stuff out of our cart if the cart is where the item is

-Actually abort activity during various failure states

RenechCDDA committed:
Fix infinite loop when sorting with no space
